sycl/include/sycl/atomic_ref.hpp

```diff
@@ -523,12 +526,53 @@ class atomic_ref_impl<
   using atomic_ref_base<T, DefaultOrder, DefaultScope, AddressSpace>::ptr;
 };
 
+// Partial specialization for 64-bit integral types needed for optional kernel
+// features
+template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
+          access::address_space AddressSpace>
+#ifndef __SYCL_DEVICE_ONLY__
+class atomic_ref_impl<
+#else
+class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
+#endif
+    T, /*IsAspectAtomic64AttrUsed = */ true, DefaultOrder, DefaultScope,
+    AddressSpace, typename detail::enable_if_t<std::is_integral<T>::value>>
+    : public atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false,
+                             DefaultOrder, DefaultScope, AddressSpace> {
+public:
+  using atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false, DefaultOrder,
+                        DefaultScope, AddressSpace>::atomic_ref_impl;
+  using atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false, DefaultOrder,
+                        DefaultScope, AddressSpace>::atomic_ref_impl::operator=;
+};
+
+// Partial specialization for 64-bit floating-point types needed for optional
+// kernel features
+template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
+          access::address_space AddressSpace>
+#ifndef __SYCL_DEVICE_ONLY__
+class atomic_ref_impl<
+#else
+class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
+#endif
+    T, /*IsAspectAtomic64AttrUsed = */ true, DefaultOrder, DefaultScope,
+    AddressSpace,
+    typename detail::enable_if_t<std::is_floating_point<T>::value>>
+    : public atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false,
+                             DefaultOrder, DefaultScope, AddressSpace> {
+public:
+  using atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false, DefaultOrder,
+                        DefaultScope, AddressSpace>::atomic_ref_impl;
+  using atomic_ref_impl<T, /*IsAspectAtomic64AttrUsed = */ false, DefaultOrder,
+                        DefaultScope, AddressSpace>::atomic_ref_impl::operator=;
+};
+
```

Comment:
Probably I am missing something but isn't it possible to merge these two specializations into one with detail::enable_if_t<std::is_integral<T>::value>|| std:is_floating_point<T>::value>> ? They seem to be similar.

Reply:
With this, tests fail like this:

```
[0/2] Re-checking globbed directories...
[0/2] Running LLVM lit test suite
-- Testing: 10 tests, 10 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..

Testing Time: 0.05s
  Passed: 10
[1/2] Running device-agnostic SYCL regression tests for SPIR-V
llvm-lit: /localdisk2/dvodopya/dv-llvm/sycl/test/lit.cfg.py:102: note: Triple: spir64-unknown-unknown
llvm-lit: /localdisk2/dvodopya/dv-llvm/llvm/utils/lit/lit/llvm/config.py:459: note: using clang: /localdisk2/dvodopya/dv-llvm/build/bin/clang
llvm-lit: /localdisk2/dvodopya/dv-llvm/sycl/test/Unit/lit.cfg.py:80: note: Using Mock Plugin.
llvm-lit: /localdisk2/dvodopya/dv-llvm/sycl/test/Unit/lit.cfg.py:83: note: SYCL cache directory: /localdisk2/dvodopya/dv-llvm/build/sycl_cache
-- Testing: 461 tests, 12 workers --
Testing: 
FAIL: SYCL :: basic_tests/atomic-ref-instantiation.cpp (1 of 461)
******************** TEST 'SYCL :: basic_tests/atomic-ref-instantiation.cpp' FAILED ********************
Script:
--
: 'RUN: at line 1';   /localdisk2/dvodopya/dv-llvm/build/bin/clang --driver-mode=g++  -fsycl -fsyntax-only -Xclang -verify /localdisk2/dvodopya/dv-llvm/sycl/test/basic_tests/atomic-ref-instantiation.cpp -Xclang -verify-ignore-unexpected=note
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 1"
$ "/localdisk2/dvodopya/dv-llvm/build/bin/clang" "--driver-mode=g++" "-fsycl" "-fsyntax-only" "-Xclang" "-verify" "/localdisk2/dvodopya/dv-llvm/sycl/test/basic_tests/atomic-ref-instantiation.cpp" "-Xclang" "-verify-ignore-unexpected=note"
# command stderr:
error: 'error' diagnostics seen but not expected: 
  File /localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp Line 696: ambiguous partial specializations of 'atomic_ref_impl<double, true, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>'
1 error generated.

error: command failed with exit status: 1

--

********************
Testing: 
FAIL: SYCL :: optional_kernel_features/atomic_ref-atomic64-aspect.cpp (2 of 461)
******************** TEST 'SYCL :: optional_kernel_features/atomic_ref-atomic64-aspect.cpp' FAILED ********************
Script:
--
: 'RUN: at line 1';   /localdisk2/dvodopya/dv-llvm/build/bin/clang --driver-mode=g++  /localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp -S -o /localdisk2/dvodopya/dv-llvm/build/tools/sycl/test/optional_kernel_features/Output/atomic_ref-atomic64-aspect.cpp.tmp.ll -fsycl-device-only
: 'RUN: at line 2';   FileCheck /localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp --input-file /localdisk2/dvodopya/dv-llvm/build/tools/sycl/test/optional_kernel_features/Output/atomic_ref-atomic64-aspect.cpp.tmp.ll
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 1"
$ "/localdisk2/dvodopya/dv-llvm/build/bin/clang" "--driver-mode=g++" "/localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp" "-S" "-o" "/localdisk2/dvodopya/dv-llvm/build/tools/sycl/test/optional_kernel_features/Output/atomic_ref-atomic64-aspect.cpp.tmp.ll" "-fsycl-device-only"
# command stderr:
In file included from /localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:10:
In file included from /localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/sycl.hpp:15:
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:696:14: error: ambiguous partial specializations of 'atomic_ref_impl<double, true, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>'
    : public detail::atomic_ref_impl<T, sizeof(T) == 8, DefaultOrder,
             ^
/localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:19:11: note: in instantiation of template class 'sycl::atomic_ref<double, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>' requested here
          sycl::atomic_ref<double, sycl::memory_order_acq_rel,
          ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:424:7: note: partial specialization matches [with T = double, IsAspectAtomic64AttrUsed = true, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class atomic_ref_impl<
      ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:536:63: note: partial specialization matches [with T = double, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
                                                              ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:696:14: error: ambiguous partial specializations of 'atomic_ref_impl<long long, true, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>'
    : public detail::atomic_ref_impl<T, sizeof(T) == 8, DefaultOrder,
             ^
/localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:26:11: note: in instantiation of template class 'sycl::atomic_ref<long long, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>' requested here
          sycl::atomic_ref<long long, sycl::memory_order_acq_rel,
          ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:276:7: note: partial specialization matches [with T = long long, IsAspectAtomic64AttrUsed = true, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class atomic_ref_impl<T, IsAspectAtomic64AttrUsed, DefaultOrder, DefaultScope,
      ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:536:63: note: partial specialization matches [with T = long long, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
                                                              ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:696:14: error: ambiguous partial specializations of 'atomic_ref_impl<unsigned long, true, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>'
    : public detail::atomic_ref_impl<T, sizeof(T) == 8, DefaultOrder,
             ^
/localdisk2/dvodopya/dv-llvm/sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:33:11: note: in instantiation of template class 'sycl::atomic_ref<unsigned long, sycl::memory_order::acq_rel, sycl::memory_scope::device, sycl::access::address_space::local_space>' requested here
          sycl::atomic_ref<uint64_t, sycl::memory_order_acq_rel,
          ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:276:7: note: partial specialization matches [with T = unsigned long, IsAspectAtomic64AttrUsed = true, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class atomic_ref_impl<T, IsAspectAtomic64AttrUsed, DefaultOrder, DefaultScope,
      ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:536:63: note: partial specialization matches [with T = unsigned long, DefaultOrder = sycl::memory_order::acq_rel, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::local_space]
class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
                                                              ^
3 errors generated.

error: command failed with exit status: 1

--

********************
Testing: 
FAIL: SYCL :: regression/reduction_kernel_conflict.cpp (3 of 461)
******************** TEST 'SYCL :: regression/reduction_kernel_conflict.cpp' FAILED ********************
Script:
--
: 'RUN: at line 1';   /localdisk2/dvodopya/dv-llvm/build/bin/clang --driver-mode=g++  -fsycl -fsyntax-only -sycl-std=2020 /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 1"
$ "/localdisk2/dvodopya/dv-llvm/build/bin/clang" "--driver-mode=g++" "-fsycl" "-fsyntax-only" "-sycl-std=2020" "/localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp"
# command stderr:
In file included from /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:6:
In file included from /localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/sycl.hpp:15:
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:696:14: error: ambiguous partial specializations of 'atomic_ref_impl<double, true, sycl::memory_order::relaxed, sycl::memory_scope::device, sycl::access::address_space::global_space>'
    : public detail::atomic_ref_impl<T, sizeof(T) == 8, DefaultOrder,
             ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:238:24: note: in instantiation of template class 'sycl::atomic_ref<double, sycl::memory_order::relaxed, sycl::memory_scope::device, sycl::access::address_space::global_space>' requested here
      auto AtomicRef = sycl::atomic_ref<T, memory_order::relaxed,
                       ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:260:5: note: in instantiation of function template specialization 'sycl::detail::combiner<sycl::reducer<double, std::plus<>, 0, 1>>::atomic_combine_impl<sycl::access::address_space::global_space, double, (lambda at /localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:261:21)>' requested here
    atomic_combine_impl<Space>(
    ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:1152:19: note: in instantiation of function template specialization 'sycl::detail::combiner<sycl::reducer<double, std::plus<>, 0, 1>>::atomic_combine<sycl::access::address_space::global_space, double, std::plus<>>' requested here
          Reducer.atomic_combine(&Out[0]);
                  ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:1138:19: note: in instantiation of function template specialization 'sycl::detail::reduction_impl_algo<double, std::plus<>, 0, 1, double *>::withInitializedMem<sycl::detail::reduction::MainKrn<kernel, sycl::detail::reduction::strategy::group_reduce_and_atomic_cross_wg>, (lambda at /localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:1138:49)>' requested here
    Redu.template withInitializedMem<Name>(CGH, [&](auto Out) {
                  ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:2164:21: note: in instantiation of function template specialization 'sycl::detail::NDRangeReduction<sycl::detail::reduction::strategy::group_reduce_and_atomic_cross_wg>::run<kernel, 1, sycl::ext::oneapi::experimental::properties<std::tuple<>>, (lambda at /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:15:10), sycl::detail::reduction_impl<double, std::plus<>, 0, 1, double *>>' requested here
      Impl.template run<KernelName>(CGH, Queue, NDRange, Properties, Redu,
                    ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/reduction.hpp:2205:40: note: in instantiation of function template specialization 'sycl::detail::NDRangeReduction<sycl::detail::reduction::strategy::auto_select>::run<kernel, 1, sycl::ext::oneapi::experimental::properties<std::tuple<>>, (lambda at /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:15:10), sycl::detail::reduction_impl<double, std::plus<>, 0, 1, double *>>' requested here
  NDRangeReduction<Strategy>::template run<KernelName>(CGH, CGH.MQueue, NDRange,
                                       ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/handler.hpp:2058:13: note: in instantiation of function template specialization 'sycl::detail::reduction_parallel_for<kernel, sycl::detail::reduction::strategy::auto_select, 1, sycl::ext::oneapi::experimental::properties<std::tuple<>>, sycl::detail::reduction_impl<double, std::plus<>, 0, 1, double *>, (lambda at /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:15:10)>' requested here
    detail::reduction_parallel_for<KernelName>(*this, Range, Properties,
            ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/handler.hpp:2066:5: note: in instantiation of function template specialization 'sycl::handler::parallel_for<kernel, 1, sycl::ext::oneapi::experimental::properties<std::tuple<>>, sycl::detail::reduction_impl<double, std::plus<>, 0, 1, double *> &, (lambda at /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:15:10)>' requested here
    parallel_for<KernelName>(
    ^
/localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:13:10: note: in instantiation of function template specialization 'sycl::handler::parallel_for<kernel, 1, sycl::detail::reduction_impl<double, std::plus<>, 0, 1, double *> &, (lambda at /localdisk2/dvodopya/dv-llvm/sycl/test/regression/reduction_kernel_conflict.cpp:15:10)>' requested here
     CGH.parallel_for<class kernel>(
         ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:424:7: note: partial specialization matches [with T = double, IsAspectAtomic64AttrUsed = true, DefaultOrder = sycl::memory_order::relaxed, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::global_space]
class atomic_ref_impl<
      ^
/localdisk2/dvodopya/dv-llvm/build/bin/../include/sycl/atomic_ref.hpp:536:63: note: partial specialization matches [with T = double, DefaultOrder = sycl::memory_order::relaxed, DefaultScope = sycl::memory_scope::device, AddressSpace = sycl::access::address_space::global_space]
class [[__sycl_detail__::__uses_aspects__(aspect::atomic64)]] atomic_ref_impl<
                                                              ^
1 error generated.

error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..
********************
Failed Tests (3):
  SYCL :: basic_tests/atomic-ref-instantiation.cpp
  SYCL :: optional_kernel_features/atomic_ref-atomic64-aspect.cpp
  SYCL :: regression/reduction_kernel_conflict.cpp
```

Reply:
I see, thank you for trying.